File: backend/src/domain/value-objects/email.ts
export class Email {
  private readonly _value: string;

  private static readonly EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

  private constructor(value: string) {
    this._value = value;
  }

  static create(value: string): Email {
    if (!value || value.trim() === '') {
      throw new Error('Email cannot be empty');
    }
    if (!Email.EMAIL_REGEX.test(value)) {
      throw new Error('Invalid email format');
    }
    return new Email(value);
  }

  get value(): string {
    return this._value;
  }

  equals(other: Email): boolean {
    return this._value === other._value;
  }
}

File: backend-go/internal/domain/user/email.go
package user

import (
	"regexp"
	"strings"

	"github.com/posiposi/dragons-counter/backend-go/internal/domain"
)

var emailRegex = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

// Email is a value object that validates and holds an email address.
type Email struct {
	value string
}

// NewEmail creates an Email after validating the format.
func NewEmail(value string) (Email, error) {
	if strings.TrimSpace(value) == "" {
		return Email{}, domain.NewError("INVALID_EMAIL", "Email cannot be empty")
	}
	if !emailRegex.MatchString(value) {
		return Email{}, domain.NewError("INVALID_EMAIL", "Invalid email format")
	}
	return Email{value: value}, nil
}

func (e Email) Value() string {
	return e.value
}

func (e Email) Equals(other Email) bool {
	return e.value == other.value
}
